compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1505,12 +1505,33 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
             false
         }
 
+        // This & will try to preserve the FromJavaObjects type in upper bounds
+        // For example, (? <: FromJavaObjects | Null) & (? <: Any),
+        // we want to get (? <: FromJavaObjects | Null) intead of (? <: Any),
+        // because we may check FromJavaObjects | Null <:< Object | Null later.
+        def bounds_&(tp1: TypeBounds, tp2: TypeBounds) =
```

Comment:
While this works, I'm afraid the same issue could appear when using `&` somewhere else, have you tried adding this logic to TypeComparer#glb instead (which is what `&` will end up calling) ?